### ISSUE TITLE
no such file or directory - .clasp.json

### ISSUE THREAD

**Issue by nazincubeta** [google-marketing-solutions/vigenair#3]
So I'm trying to deploy using Windows via Windows Subsystem for Linux 2 terminal...

The installation breaks and aborts with the following error:

````
Creating Apps Script Project...
node:internal/process/promises:288
            triggerUncaughtException(err, true /* fromPromise */);
            ^

[Error: ENOENT: no such file or directory, lstat 'C:\github\google-marketing-solutions\vigenair\dist\.clasp.json'] {
  errno: -4058,
  code: 'ENOENT',
  syscall: 'lstat',
  path: 'C:\\github\\google-marketing-solutions\\vigenair\\dist\\.clasp.json'
}
````

I've installed Google Clasp via NPM beforehand.

**Comment by nazincubeta** [comment]
I've also tried from Windows Terminal + Git Bash to run the deploy.sh file, same result as above.

**Comment by mohabfekry** [comment]
deploy.sh is rather for deploying the backend cloud functions, it definitely won't throw this error :)

WIth regards to your clasp installation, please make sure it is installed **globally**
`npm install @google/clasp -g`

Installing it without the `-g` flag will not suffice.

**Comment by nazincubeta** [comment]
Yes, it's been installed this way: ```npm install @google/clasp -g``` and I'm still getting the error.

**Comment by nazincubeta** [comment]
What I'm running is ```npm start``` and somewhere down the line, it throws that error immediately after it says:

```state: ACTIVE
updateTime: '2024-06-27T17:52:31.332998315Z'
url: https://someurl/vigenair```

**Comment by nazincubeta** [comment]
It seems to break somewhere in Service, right after Build.

```
INFO - Deploying the 'vigenair' Cloud Function...
Preparing function...done.
X  Updating function (may take a while)...
  OK [Build] Logs are available at [https://console.cloud.google.com/cloud-build/builds;region=redactedforsecurity]
     [Service]
  .  [Trigger]
  .  [ArtifactRegistry]
  .  [Healthcheck]
  .  [Triggercheck]
Completed with warnings:
  [INFO] A new revision will be deployed serving with 100% traffic.
You can view your function in the Cloud Console here: https://console.cloud.google.com/functions/details/redactedforsecurity
```

**Comment by mohabfekry** [comment]
So the npm issue you originally raised is now resolved as expected. It seems now that your Cloud Functions deployment is not going through. Could you please check the Cloud Build logs and report the error here so that I can assist you further?

Either from here:
`https://console.cloud.google.com/functions/details/<your-cloud-function-region>/vigenair?env=gen2&project=<your-gcp-project>&tab=logs`

or most likely you'll find the issue here:
`https://console.cloud.google.com/cloud-build/builds;region=<your-cloud-function-region>?project=<your-gcp-project>`

**Comment by nazincubeta** [comment]
We seem to have no errors in the Cloud Build logs, just "Notices", with stuff like:

```
{
  "protoPayload": {
    "@type": "type.googleapis.com/google.cloud.audit.AuditLog",
    "authenticationInfo": {
      "principalEmail": "redacted@gcf-admin-robot.iam.gserviceaccount.com"
    },
    "requestMetadata": {
      "requestAttributes": {
        "time": "2024-06-28T09:04:11.216600Z",
        "auth": {}
      },
      "destinationAttributes": {}
    },
    "authorizationInfo": [
      {
        "resource": "namespaces/redacted/services/vigenair",
        "permission": "run.services.update",
        "granted": true,
        "resourceAttributes": {},
        "permissionType": "ADMIN_WRITE"
      }
    ],
    "resourceName": "namespaces/redacted/services/vigenair",
    "resourceLocation": {
      "currentLocations": [
        "redacted"
      ]
    }
  },
  "insertId": "redacted",
  "resource": {
    "type": "cloud_run_revision",
    "labels": {
      "location": "redacted",
      "service_name": "vigenair",
      "project_id": "us-tech-in-vigenair",
      "configuration_name": "",
      "revision_name": ""
    }
  },
  "timestamp": "2024-06-28T09:04:11.048626Z",
  "severity": "NOTICE",
  "logName": "projects/redacted/logs/cloudaudit.googleapis.com%2Factivity",
  "receiveTimestamp": "2024-06-28T09:04:11.793072275Z"
}
```

**Comment by nazincubeta** [comment]
And no, the issue I originally raised is not resolved, I'm still getting the error message, etc in Terminal when I run npm start as in my 1st post. Thanks.

**Comment by nazincubeta** [comment]
Has install/deployment of Vigenair of this been tested on Windows or just Macs? I had to change line 25 in package json from:

```
   "deploy-service": "cd service && ./deploy.sh",
```

to

```
   "deploy-service": "cd service && sh deploy.sh",
```

and install Git Bash to even get the installation process to progress to where I have.

running ```./deploy.sh``` in terminal only opens the file in a text editor and does not run it in Windows without Git Bash or similar SH programs installed.

**Comment by mohabfekry** [comment]
Alright one issue after the other.

The original npm issue you raised was related to deploying the frontend (not cloud functions and not deploy.sh). That should be resolved with the global installation of `clasp`. Just please make sure you delete any remnants of old broken files from your repo:

<vigenair_root>/dist/.clasp.json
<vigenair_root>/ui/.clasp.json
<vigenair_root>/ui/.clasp-dev.json
<vigenair_root>/ui/.clasp-prod.json

Before re-running `npm start`. Please do that, deploy the frontend and confirm that the issue is fixed.

Now you're raising another issue that you're having trouble deploying the backend using `npm start`. If you navigate to `<vigenair_root>/service/` are you able to run deploy.sh directly? Does it work then or fail?

The `npm start` script for the backend deployment more or less just calls `deploy.sh`, so this might help you carry on until we figure out what's the problem with windows and the "." notation of shell. We might just indeed have to prefix the call with `sh` or `bash` (depending on whatever is more universally supported).

**Comment by nazincubeta** [comment]
It seems that no .clasp.json files are ever created, I look into the directories above, and there's no trace, even during the ```npm start``` process.

**Comment by nazincubeta** [comment]
doing sh deploy.sh completes, and ends with a URL. Going to that URL says:

```
Error: Forbidden
Your client does not have permission to get URL /vigenair from this server.
```

**Comment by mohabfekry** [comment]
That's correct, you shouldn't access the cloud function via web as it is triggered via Cloud Storage. The URL of the app is what you receive when you deploy the frontend.

**Comment by nazincubeta** [comment]
The original issue isn't resolved. Clasp is properly installed globally, there are no .clas.json files ever created that I see (unless they are supposed to be created and deleted fast by the process), and I'm still getting the error message in my original post at the very top of this thread. I don't think I've ever seen instructions during the ```npm start``` process on how to access Vigenair.

**Comment by mohabfekry** [comment]
What do you see when you run "which clasp" in the command line ( I think it's "where.exe" on windows)?

If so are you able to execute this on the command line in <vigenair_root>:

`npx clasp create --type sheets --rootDir ./dist --title ViGenAiR`

If that completes successfully you should find a `.clasp.json` file and an `appsscript.json` file under <vigenair_root>/dist. Delete the `appsscript.json` file and move the `.clasp.json` file under <vigenair_root>/ui and rename it to `.clasp-dev.json`.

Once this is done you should be able to "deploy" the application by navigation to <vigenair_root>/ui and running `npm run deploy:dev`. This will output the ID of the web app deployment, which you should then place in the URL below to access the app:

https://script.google.com/a/macros/google.com/s/<webAppDeploymentId>/exec

These are the steps that the `npm start` script does but since you're running into issues with that I've asked you to perform the steps manually so we can identify where the problem is.

**Comment by mohabfekry** [comment]
What do you see when you run "which clasp" in the command line ( I think it's "where.exe" on windows)?

If so are you able to execute this on the command line in <vigenair_root>:

`npx clasp create --type sheets --rootDir ./dist --title ViGenAiR`

If that completes successfully you should find a `.clasp.json` file and an `appsscript.json` file under <vigenair_root>/dist. Delete the `appsscript.json` file and move the `.clasp.json` file under <vigenair_root>/ui and rename it to `.clasp-dev.json`.

Once this is done you should be able to "deploy" the application by navigation to <vigenair_root>/ui and running `npm run deploy:dev`. This will output the ID of the web app deployment, which you should then place in the URL below to access the app:

https://script.google.com/a/macros/google.com/s/<webAppDeploymentId>/exec

These are the steps that the `npm start` script does but since you're running into issues with that I've asked you to perform the steps manually so we can identify where the problem is.

**Comment by nazincubeta** [comment]
Thank you for your patience!

Clasp is in my nodejs installation folder.

After following the steps outlined in your last post above, then running ```npm start``` again, the process properly deployed, and it gave me a URL with which I could access the Vigenair UI.

I'm currently testing it out, loading a 2 minute and a 1 minute video clip with "Analyse voice-over" checked ended up with an error message asking me to start over.

A 30-second clip with "Analyse voice-over" unchecked succeeded.

Are there duration limits to video clip length that Vigenair can ingest?

Finally, has the Github repository and deployment instructions been tested against Windows and not just Mac/Linux? Just curious as the issues I've run into seem to be Windows-specific. Tweaking the repo code for better Windows compatibility would help a lot :)

**Comment by nazincubeta** [comment]
Another question is that would the deployment region matter in terms of performance? If I remember correctly certain datacenters have different GPU setups. I set up our test deployment at the default us-central1.

**Comment by mohabfekry** [comment]
There are no duration or size limits beyond what normal browsers support (which are files up to 2GB-4GB), but your input regarding the "analyse voice-over" checkbox pointed me to an issue which I found and fixed.

The cloud functions environment was updated recently by Google Cloud and there now seems to be a "protobuf" version issue with the dependencies of ViGenAiR (you'll find it if you check the cloud function logs), and it gets triggered when we're trying to analyse the voice-over on a video. We actually don't use protobuf directly but some google-libs we depend on do., so it affects us nevertheless. I've fixed this now with: https://github.com/google-marketing-solutions/vigenair/commit/7709ca7e3b5630bd18d1a6837bb7606cbfe71ebe.

Indeed we've only tested the installation steps on unix-based systems, so we'll make sure we add more windows- / universally-friendly instructions in the future. 

Thanks again for your patience and the report!
